crates/astria-sequencer/src/service/mempool/mod.rs

```diff
@@ -120,50 +127,181 @@ async fn handle_check_tx<S: accounts::StateReadExt + address::StateReadExt + 'st
 ) -> response::CheckTx {
     use sha2::Digest as _;
 
-    let start_parsing = Instant::now();
-
     let request::CheckTx {
         tx, ..
     } = req;
 
     let tx_hash = sha2::Sha256::digest(&tx).into();
+
+    // check if the transaction has been removed from the appside mempool
+    if let Some(rsp) = check_removed_comet_bft(tx_hash, mempool, metrics).await {
+        return rsp;
+    }
+
+    // check if the transaction is already in the mempool
+    if let Some(rsp) = check_tracked(tx_hash, mempool, metrics).await {
+        return rsp;
+    }
+
+    // perform stateless checks
+    let signed_tx = match stateless_checks(tx, &state, metrics).await {
+        StatelessReturn::Response(rsp) => return rsp,
+        StatelessReturn::Tx(signed_tx) => signed_tx,
+    };
+
+    // attempt to insert the transaction into the mempool
+    if let Some(rsp) = insert_into_mempool(mempool, &state, signed_tx, metrics).await {
+        return rsp;
+    }
```

Comment:
this function returning an `Option` that's only set if there's an error response is somewhat confusing, would prefer for this function to return a `Result<(), response::CheckTx>` so the error variant is the error response

Reply:
same for `check_removed_comet_bft` and `check_tracked`

Reply:
Changed it, feels much better that way